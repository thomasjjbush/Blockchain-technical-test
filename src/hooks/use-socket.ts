import { useEffect, useRef } from 'react';
import { useDispatch } from 'react-redux';
import { SocketOptions } from '../../types';

export const useSocket = <T>(url: string, options?: SocketOptions<T>): void => {
    const { current: socket } = useRef(new WebSocket(url));
    const dispatch = useDispatch();

    useEffect(() => {
        socket.onclose = () => socket.send(JSON.stringify({ op: options.onClose }));
        socket.onerror = () => dispatch({ type: options.onError });
        socket.onopen = () => socket.send(JSON.stringify({ op: options.onOpen }));
        socket.onmessage = (e) => dispatch({ type: options.onMessage, payload: JSON.parse(e.data).x });

        return () => socket.close();
    }, []);
};
