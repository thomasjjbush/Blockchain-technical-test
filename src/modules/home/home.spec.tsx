import React from 'react';
import { shallow } from 'enzyme';
import { useSocket } from '../../hooks/use-socket';
import Home from './home';

jest.mock('react-redux', () => ({
    useSelector: jest.fn().mockReturnValue({ latest_transactions: [{ hash: 'hash', size: 20 }] }),
}));

jest.mock('../../hooks/use-socket');

describe('Home', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('should match snapshot', () => {
        expect(shallow(<Home />)).toMatchSnapshot();
    });

    it('should invoke useSocket hook with correct arguments', () => {
        shallow(<Home />);
        expect(useSocket).toHaveBeenCalledTimes(1);
        expect(useSocket).toHaveBeenCalledWith('wss://ws.blockchain.info/inv', {
            onClose: 'unconfirmed_unsub',
            onError: 'home/ERROR',
            onMessage: 'home/LATEST_TRANSACTION',
            onOpen: 'unconfirmed_sub',
        });
    });
});
