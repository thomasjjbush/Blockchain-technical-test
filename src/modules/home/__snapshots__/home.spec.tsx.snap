// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should match snapshot 1`] = `
<styled.div
  styles="
    background-color: white;
    padding: 30px;
    margin: 30px;
"
>
  <styled.h1
    styles="
    color: theme.colors.brand;
"
  >
    Latest unconfirmed transactions
  </styled.h1>
  <Transaction
    hash="hash"
    key="hash"
    size={20}
  />
</styled.div>
`;
